src/host-config/proxy.ts

```diff
@@ -0,0 +1,275 @@
+import log from '../lib/supervisor-console';
+import * as dbus from '../lib/dbus';
+
+const proxyBasePath = pathOnBoot('system-proxy');
+const noProxyPath = path.join(proxyBasePath, 'no_proxy');
+const redsocksConfPath = path.join(proxyBasePath, 'redsocks.conf');
+
+export const disallowedProxyFields = ['local_ip', 'local_port'];
+export const validRedsocksProxyTypes = [
+	'socks4',
+	'socks5',
+	'http-connect',
+	'http-relay',
+];
+
+const isAllowedProxyField = (field: string): boolean =>
+	['type', 'ip', 'port', 'login', 'password'].includes(field);
+
+const isAuthField = (field: string): boolean =>
+	['login', 'password'].includes(field);
+
+// ? is a lazy operator, so only the contents up until the first `}(?=\s|$)` is matched.
+// (?=\s|$) indicates that `}` must be followed by a whitespace or end of file to match,
+// in case there are user fields with brackets such as login or password fields.
+const blockRegexFor = (blockLabel: string) =>
+	new RegExp(`${blockLabel}\\s?{([\\s\\S]+?)}(?=\\s|$)`);
+
+const baseBlock = {
+	log_debug: 'off',
+	log_info: 'on',
+	log: 'stderr',
+	daemon: 'off',
+	redirector: 'iptables',
+};
+
+export class RedsocksConf {
+	public static stringify(config: RedsocksConfig): string {
+		const blocks: string[] = [];
+
+		if (config.redsocks && Object.keys(config.redsocks).length > 0) {
+			blocks.push(RedsocksConf.stringifyBlock('base', baseBlock));
+			blocks.push(
+				RedsocksConf.stringifyBlock('redsocks', {
+					...config.redsocks,
+					local_ip: '127.0.0.1',
+					local_port: 12345,
+				}),
+			);
+		}
+
+		return blocks.length ? blocks.join('\n') : '';
+	}
+
+	public static parse(rawConf: string): RedsocksConfig {
+		const conf: RedsocksConfig = {};
+		if (rawConf.length === 0) {
+			return conf;
+		}
+
+		// Extract contents of `redsocks {...}` using regex
+		const rawRedsocksBlockMatch = rawConf.match(blockRegexFor('redsocks'));
+		// If match length shorter than 2, no group was captured, indicating malformed config
+		if (!rawRedsocksBlockMatch || rawRedsocksBlockMatch.length < 2) {
+			log.warn('Invalid redsocks block in redsocks.conf');
+			return conf;
+		}
+		const rawRedsocksBlock = RedsocksConf.parseBlock(
+			rawRedsocksBlockMatch[1],
+			disallowedProxyFields,
+		);
+		const maybeProxyConfig = ProxyConfig.decode(rawRedsocksBlock);
+		if (isRight(maybeProxyConfig)) {
+			conf.redsocks = {
+				...maybeProxyConfig.right,
+			};
+		} else {
+			log.warn(Reporter.report(maybeProxyConfig).join('\n'));
+			// Allow the invalid values for known fields through for
+			// backwards compatibility, while excluding unknown fields.
+			const maybeInvalidProxyConfig: Record<string, string> = {};
+			for (const key in rawRedsocksBlock) {
+				if (isAllowedProxyField(key)) {
+					maybeInvalidProxyConfig[key] = rawRedsocksBlock[key];
+				}
+			}
+
+			if (Object.keys(maybeInvalidProxyConfig).length) {
+				conf.redsocks = maybeInvalidProxyConfig as unknown as ProxyConfig;
```

Comment:
I feel like this violates the contract defined by the function signature. The function declares that it returns a `RedSocks` config but it could technically return an object with a missing `type` or `ip`. We could either return a default value if `.conf` does not include a property or return an empty config if none of the required values are defined.

Reply:
I see what you mean here. Between the two suggestions, maybe warning in the console and returning nothing is better rather than returning a default value. However, given there is already a check which filters for accepted `redsocks` block keys above this `as unknown as ProxyConfig` cast:
```
for (const key in rawRedsocksBlock) {
  if (isAllowedProxyField(key)) {
    maybeInvalidProxyConfig[key] = rawRedsocksBlock[key];
  }
}
```
I think it's okay and maybe even necessary for backwards compatibility, although unpalatable -- and I do agree that it's unpalatable.

Reply:
I don't think it's just ugly, I think something like this is potentially dangerous. When you have a function signature 

```
public static parse(rawConf: string): RedsocksConfig
```

You are ensuring whoever uses the function that the result is a valid object of type `RedsocksConfig`, which is not true with this implementation.

I think if the configuration on disk is invalid we should just ignore it and treat it the same as if there was no configuration at all.

Regarding backwards compatibility, I don't think this breaks it. If `redsocks.conf` is broken (e.g. it is missing a required field), then the device proxy would not be functional anyway, so there is no difference if we treat the invalid configuration as no configuration

Reply:
Noted and understood, thanks for explaining. I've applied the change suggested [here](https://github.com/balena-os/balena-supervisor/pull/2329/commits/1e224be0cd167d7f037a3d1cb4b260b6ff36d458#diff-87da4178ce07f20bae291890a95e15b8bf7990d89aeb29ec6bfe1f77b508c3e5R56-R63) of returning no configuration if a valid config cannot be parsed from redsocks.conf.